jetstream_pt/third_party/gemma/model.py

```diff
@@ -35,6 +35,142 @@ def precompute_freqs_cis(
   return freqs_cis
 
 
+def reshape_for_broadcast(
+    freqs_cis: torch.Tensor, x: torch.Tensor
+) -> torch.Tensor:
+  ndim = x.ndim
+  assert 1 < ndim
+  assert freqs_cis.shape == (
+      x.shape[0],
+      x.shape[2],
+      x.shape[3],
+  ), f"freqs_cis: {freqs_cis.shape }, x: {x.shape}"
+  shape = [d if i != 1 else 1 for i, d in enumerate(x.shape)]
+  return freqs_cis.view(*shape)
+
+
+def apply_rotary_emb(x: torch.Tensor, freqs_cis: torch.Tensor) -> torch.Tensor:
+  """Applies the rotary embedding to the query and key tensors."""
+  x_ = torch.view_as_complex(
+      torch.stack(torch.chunk(x.transpose(1, 2).float(), 2, dim=-1), dim=-1)
+  )
+  freqs_cis = reshape_for_broadcast(freqs_cis, x_)
+  x_out = torch.view_as_real(x_ * freqs_cis).type_as(x)
+  x_out = torch.cat(torch.chunk(x_out, 2, dim=-1), dim=-2)
+  x_out = x_out.reshape(
+      x_out.shape[0], x_out.shape[1], x_out.shape[2], -1
+  ).transpose(1, 2)
+  return x_out
+
+
+class GemmaAttention(nn.Module):
```

Comment:
In long term, we might need to extend from Attention class, large percentage of code are similar.